Makefile

ProkopRandacek opened the pull request:
Hello,
I cleaned up the makefile a little bit.

- The output when compiling is now nicely formatted
- Created CFLAGS that are common for static and dynamic builds
- .c files are now generated from the src directory content using make functions
- renamed all variables to uppercase for consistency
- removed unused variables
- `static` and `dynamic` rules for building only one type of library

I tested the Makefile on Linux. I can't test it on windows but I'm pretty sure it will work everywhere.
I also wanted to ask if the `build_linux.sh` and `build_windows_*.bat` scripts are used or if it is safe to perhaps delete them?

At least on my machine the `build_linux.sh` script fails to compile.

And I also compressed the pngs because it makes me happy :)

vtereshkov commented:
@ProkopRandacek Thank you!
- The Makefile itself is fine, except that I kindly ask you to do something with the `src/*.o` files: either keep them in a separate folder like `obj`, or move `*_static.o` to `build` (since they can be necessary for static linking) and completely remove `*_dynamic.o`.
- The `build_linux.sh` and `build_windows_*.bat` scripts are crucial for me. Actually this is the main tool I usually use for building and packaging Umka. This is particularly important on Windows, since the `make` tool is not standard there. Moreover, the Windows Makefile (I mean the standalone MinGW, not MSYS2) would be completely different from what you have prepared (see, for example, the command-line arguments for building a DLL in `build_windows_mingw.bat`.
- Please don't add `src/umka_runtime_src.h` to `.gitignore`.

ProkopRandacek commented:
- Moved `src/*.o` to `obj/*.o`
- Removed dynamic building rules

Yes the `build_linux.sh` script was failing because of the `src/*.o` files left by makefile. Now that they are in another folder this can't happen.

Why can't be `src/umka_runtime_src.h` in the `.gitignore`? It is copied at build from `import_embed/umka_runtime_src.h` (and as a side effect of this makes git think that the repo is dirty after every build which I find annoying).

ProkopRandacek commented:
PS:
Speaking of `src/umka_runtime_src.h`, It came to my mind that it could be generated directly from the `import_embed/std.um` file by some rule in the Makefile. That would eliminate the need to use the `um2h.um` script which would be also quite nice imo.
But i don't think it can be done only with Makefile's functions and thus possibly breaking the compatibility with operating systems that don't have the external tools needed for this generation (probably just `sed`).

([I do a similar thing to write text files into a header files in one of my projects](https://github.com/ProkopRandacek/r8/blob/main/src/Makefile#L19-L27))

ProkopRandacek commented:
Just thought of a feature:
Create a `include` dir in the `build` dir and put all the files that are supposed to be included there. Instead of having to include the entire `src` dir like it is right now.

vtereshkov commented:
@ProkopRandacek 
- OK for `.gitignore`.
- If you would like to have the `build/include` folder to store `umka_api.h`, please do it. But please keep in mind that all the header files contained in the `src` folder should be kept as they are. Don't touch them.
- It seems like you have removed the `.so` build rules completely. For me, the main way of using Umka is linking the `.so` library to the program executable. I meant only removing the `*_dynamic.o` files, since they are never used for linking by the end-user (he/she would use either `*_static.o` for static linking, as Marek does, or the `.so` for dynamic linking).
- Please don't use the `make` utility for generating `umka_runtime_src.h`. This would be very inconvenient for me on Windows.
- Why is the automated MacOS build failing now?

ProkopRandacek commented:
Mac had issues with the `mkdir -p` (it was thinking that `-p` is a folder, creating it and then complaining that it already exists on the next `mkdir` call). Already fixed.

I'm not sure if you can remove the `*_dynamic.o` without removing the dynamic build. You define the `-DUMKA_BUILD` for dynamic build which can result in a different `*.o` than `*.o` created by static build which defines the `-DUMKA_STATIC`. The current dynamic build uses the `*_dynamic.o`.
I can re-add the dynamic build back with the `*_dynamic.o` files.

vtereshkov commented:
@ProkopRandacek I mean you can just run something like `rm src/*_dynamic.o` after the dynamic build is complete.

ProkopRandacek commented:
Oh i see.

Why would you do that? They are in a separate directory and can be reused for the next build if the source files didn't change. Or you can compile the source code files directly into a `.so` file without the intermediate step of creating `.o` files if that's what you want.

If you want to clear the repo there is a `clean` rule that deletes all intermediate build files.
